app/services/task_service/gpt_generate_tests.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+module TaskService
+  class GptGenerateTests < ServiceBase
+    def initialize(task:)
+      super()
+      @task = task
+      @message = task.description
+      @language = @task.programming_language&.language
+      raise MissingLanguageError if @language.blank?
+    end
+
+    def execute
+      programming_language = @task.programming_language
+      file_extension = programming_language.file_extension
+
+      file_content = gpt_response
+      test_file = TaskFile.new(content: file_content, name: "test#{file_extension}", used_by_grader: true, visible: false,
+        xml_id: SecureRandom.uuid)
+      test = Test.new(task: @task, title: I18n.t('tests.model.generated_test'), xml_id: SecureRandom.uuid,
+        files: [test_file])
+
+      @task.tests << test
+    end
+
+    private
+
+    def client
+      @client ||= OpenAI::Client.new
+    end
+
+    def gpt_response
+      # train client with some prompts
+      messages = training_prompts.map do |prompt|
+        {role: 'system', content: prompt}
+      end
+
+      # send user message
+      messages << {role: 'user', content: @message}
+
+      # create gpt client
+      response = client.chat(
+        parameters: {
+          model: 'gpt-3.5-turbo',
+          messages:,
+          temperature: 0.7, # Lower values insure reproducibility
+        }
+      )
+
+      # parse out the response
+      raw_response = response.dig('choices', 0, 'message', 'content')
+
+      # check for ``` in the response and extract the text between the first set
+      raise InvalidTaskDescription unless raw_response.include?('```')
```

Comment:
with the current prompt, if the model responds correctly, then it will include the code markdown which we can then extract. However, if the task description is unclear, it will only include text. I have tried this locally a few times and this seemed to be the best way to get a consistent output from the model and detect when the description was not sufficient to yield a code output.

Reply:
Sorry, I misunderstood what you were asking before. 
The model's explanation varries everytime in term of size and content. Showing directly what the explanation of the model is, might cause more confusion sometimes. the output would need to be analysed and refined before it can be shown back to the user showing a clearer explanation and guides to improve. Another area of future work I think that can be useful.

Reply:
Okay, fine for me. Would you please mind to create a follow-up ticket where you collect all the ideas we gathered during the current implementation / review? For example, I am thinking of:
- per-user tokens for ChatGPT
- showing the model's error message to users
- improving the file name (extension + name)
- Thinking of further "promotions" (optional)
- Maybe have this view as part of the "edit" view (since this could allow edits before they are persisted). However, this also "hides" the feature more which is not good
- Check a need for rate limiting
- Improved "billing" overview (i.e., on a per-user base)
- Testing the current implementation as part of a user study
- ... some further ideas we had before 😅

Reply:
Sure I can do that. I already had a couple of notes I took earlier from the comments provided by you and Karol.